### ISSUE TITLE
Epic: Finish off merge insert functionality

### ISSUE THREAD

**Issue by westonpace** [lancedb/lance#2022]
We have some basic merge insert functionality in lance today.  Other databases have more sophisticated capabilities.  Perhaps the most extensive and well documented is the "[merge into](https://learn.microsoft.com/en-us/sql/t-sql/statements/merge-transact-sql?view=sql-server-ver16)" from SQL server.  This issue will keep track of the various things that aren't yet implemented in our version.

 - [x] #2610
 - [ ] Allow update columns to be projection statements
This will allow for things like `when_matched().update({"count": "source.count + target.count"})
 - [ ] Allow updates to be applied to "when not matched by source" rows
 This allows for updating target rows that don't match anything in the source (today we can just delete them)
 - [ ] https://github.com/lancedb/lance/issues/1983
 - [ ] https://github.com/lancedb/lance/issues/2621

**Comment by mutecamel** [comment]
`merge_insert` is one of the reasons I switch to Lance and LanceDB. Thank you for your hard work!

**Comment by valkum** [comment]
I think this is missing one important feature to be able to support proper lazy embeddings in lancedb.
Today you can't delete rows when not matched by source that do not contain the embedding.
I already pulled out the addition out of `merge_insert` (to compute embeddings lazily). Updates and deletions shouldn't really require the full schema including the embeddings. The only columns required should be the join key and the updated columns.

Currently partial schema writes will trigger fragment rewrites. This branch states that it does not support deleting rows. IIRC there are two write modes, `RewriteRows` and `RewriteColumns`. `RewriteRows` is faster for drops and additions of rows but requires the full schema. 
If lance would, in the case of partial schemas and `RewriteColumns`, rewrite all columns, it could potentially support the case of deletions, right?
To circumvent this current limitation, we would need to join prior to calling merge_insert to get the already present embeddings.

**Comment by wjones127** [comment]
*I started thinking about this with Claude, but need to put it down again. Posting the high-level plan we came up with here for future reference.*

# merge_insert on DataFusion: end-state vision + transition

Local planning doc, rev 3 (2026-07-28). Target: `upstream/main` (lance-format/lance). Not public.

**Priorities: (1) BTREE-key workloads on v2 soon, (2) delete v1 soon, (3) perf third.**

Why this rev: previous simplification attempts ADDED complexity — each built new
merge-owned machinery (probe providers, routing gates, eligibility lists) beside
the old instead of moving decisions to owners that already exist. This rev pins
the end state concretely enough to check every diff against it.
**Decided 2026-07-28: discard the branch's `IndexProbeTarget` in favor of filter
pushdown (Phase 1).**

---

## Part I — The end state

merge_insert = full SQL `MERGE INTO`. One canonical lowering; zero plan-shape
decisions in merge code. Merge owns exactly three things: **the lowering, the
clause compiler, the sinks**. Everything else is owned by DF optimizer stages,
the shared scan/filter/index stack, or sink runtime.

### 1. Code inventory at the end

```
rust/lance/src/dataset/write/merge_insert/
  mod.rs            builder + params + input validation          (exists today)
  lowering.rs       create_plan, join_type, final-value derivation   (~200 loc)
  assign_action.rs  clauses → one CASE expr                      (exists; grows arms)
  logical_plan.rs   MergeWrite node + planner (= one pure match) (exists; shrinks)
  exec/             DeleteOnly | PartialCols | FullRows sinks + shared commit/metrics

DELETED: probe.rs, Merger, create_indexed_scan_joined_stream,
  create_full_table_joined_stream, create_joined_stream, v1 update_fragments,
  can_use_create_plan, ReplayExec usage, MapIndexExec (if unconsumed)

SHARED STACKS GAIN (not merge code; benefits every query):
  lance-datafusion:  semi-join index acceleration (key filter from a join's
                     build side into a Lance scan) — any DF join over a Lance
                     table gets index acceleration, not just merge
  FilteredReadExec:  DF dynamic-filter pushdown support
  lance-index:       fast large-IsIn evaluation (sort, finger cursor, prefetch)
```

Merge-specific non-test LOC roughly halves vs today. Every phase is
net-negative in merge code except Phase 2 (ports a v1 capability; its v1 copy
dies in Phase 3).

### 2. `create_plan`, the whole thing, forever

```rust
async fn create_plan(&self, source: Arc<dyn TableProvider>) -> Result<Arc<dyn ExecutionPlan>> {
    let ctx = merge_session();                                   // lance rules registered once
    let target = ctx.read_table(lance_provider(&self.dataset))?  // plain provider, always
        .alias("target")?;
    let source = ctx.read_table(source)?
        .with_column(SENTINEL, lit(true))?
        .alias("source")?;

    let df = target
        .join(source, join_type(&self.params), &keys, &keys, None)?
        .with_column(ACTION, action_expr(&self.params, &schema)?)?;
    let df = with_final_values(df, &self.params, &schemas)?;     // SET/INSERT exprs +
                                                                 // missing-col fill, pure
    let (state, plan) = df.into_parts();
    let plan = LogicalPlan::Extension(MergeWrite::new(plan, self.dataset.clone(), self.params.clone()));
    let plan = state.optimize(&plan)?;
    physical_planner().create_physical_plan(&plan, &state).await
}
```

Notably absent: index checks, storage-version checks, source-size thresholds,
provider variants, plan-shape branches. `join_type`, `action_expr`,
`with_final_values` are total pure functions of the statement.

Transition-era exception (Phase 1 → deleted in Phase 4c): ONE extra line,
`let target = with_key_filter(target, &self.params, &source)?;` —
validity-gated static IN-list, see Part II.

### 3. The one plan shape, and who decides what

Headline upsert (small materialized source, BTREE key, `by_source=Keep`):

```
FullSchemaMergeInsertExec            sink strategy = f(shape); append-downgrade armed
└─ Projection [__action = CASE …]
   └─ HashJoin (build = source)      ← DF physical planner, from provider stats (#7368)
      ├─ MemTableScan (source)
      └─ FilteredReadExec (target)
           filter: key IN (…)        ← semi-join filter (shared rule; helper in transition)
           index: BTree(key) → exact row addrs   ← scanner's filter→index machinery
           + unindexed-fragment scan, deletion masking   ← FilteredReadExec native
```

Merge code decided: nothing.

Variations are parameter changes, not shapes:
- **Streaming source** → no static filter (dynamic min/max filter once 4c
  lands); otherwise identical. Worst case ≡ plain hash join.
- **Delete-by-source** → `join_type` derives Left/Full; key filter invalid
  (validity is structural: filter pushdown into the preserved side of an outer
  join is unsound) → not attached. Same shape.
- **Delete-only statement** → same shape, `DeleteOnlyMergeInsertExec` sink.
- **Partial-schema update** → same shape, `PartialColsMergeInsertExec` sink;
  overlay-vs-column-rewrite chosen per fragment at runtime.
- **UUIDv7 append-mostly** → same shape; page-table binary search returns no
  candidate pages (~zero IO), sink downgrades commit to Append at runtime.

### 4. The sink contract

Input stream: final column values + `target._rowaddr` + `__action`
(`Nothing | Update | Insert | Delete | Fail`). ALL clause semantics — predicates,
SET expressions, INSERT expressions, missing-column fill — are compiled into
upstream projections by the lowering. Sinks only route rows:

```
strategy = f(clauses, source schema)     one pure match, nothing else:
  DeleteOnly   matched rowaddrs → deletion files
  PartialCols  matched rows grouped per fragment → column patch | overlay
               (per-fragment runtime crossover); inserts → new fragments
  FullRows     updates+inserts → new fragments; matched rowaddrs → deletions
runtime adaptivity (sink-internal, invisible to planning):
  zero matches on pure-upsert shape → commit Operation::Append
  per-fragment overlay crossover by observed match density
```

All sinks emit full metrics (`bytes_written`, `num_files_written`,
`inserted_rows_filter`) and appear in `explain_plan`/`analyze_plan`, which work
unconditionally.

### 5. The growth test (what "full MERGE INTO" costs later)

| future feature | touches | does NOT touch |
|---|---|---|
| multiple predicated WHEN MATCHED clauses | params, `action_expr` CASE arms | plan shape, sinks, scan |
| `UPDATE SET col = expr` | params, `with_final_values` | 〃 |
| `INSERT (cols) VALUES (exprs)` | params, `with_final_values` | 〃 |
| arbitrary `ON` condition | `join_type`/predicate derivation, filter validity fn | 〃 (probe applies when equi-keys extractable) |
| `NOT MATCHED BY SOURCE THEN UPDATE` | new action variant, sink row routing | plan shape, scan |

v1 fails this test structurally (each feature = new hand-built join stream).
The end state passes it by construction: features extend pure functions.

### 6. Invariants (check every future diff against these)

1. `create_plan` = one canonical lowering. Branches on *validity* (structural
   facts) only — never on cost. Grep test: no `index`, no `use_index`, no
   thresholds in lowering.rs.
2. Access-path work lands in the scan/filter/index stack, shared with queries.
   Merge inherits it. If a change is merge-only, it's in the wrong place.
3. Cost-shaped choices live where their inputs live: DF optimizer stages
   (stats) or sink runtime (observations). Never thresholds in lowering.
4. Sink strategy = pure function of statement shape; planner = one match.
5. Conservation law: every phase net-negative in merge-specific code
   (Phase 2 excepted, and Phase 3 collects its debt).

### 7. Why the probe is not a plan shape (the reframe that pays for the rest)

The index probe = **semi-join filter pushdown**: `target ⋉ source.keys` pushed
into the target scan.

- **Static (works today)**: materialized source → `key IN (<literals>)` on the
  target DataFrame. `LanceTableProvider::supports_filters_pushdown` is already
  `Exact`; scanner maps `Expr::InList → SargableQuery::IsIn → BTreeIndex`;
  FilteredReadExec already handles unindexed fragments, deletions, legacy
  storage. Composite keys = one conjunct per column; over-selects tuples, join
  re-checks (same semantics the probe design had).
- **Dynamic (end state)**: DF 54's filter-pushdown API
  (`gather_filters_for_pushdown`/`handle_child_pushdown_result`,
  `DynamicFilterPhysicalExpr`) flows build-side keys/bounds into the probe-side
  scan after build. Lance exec nodes don't implement it yet. Covers streaming
  sources; kills plan-time key collection. Home: shared lance-datafusion rule +
  FilteredReadExec — not merge code.

Cost stops being routing: same shape either way; the filter prunes what it
prunes. Degradation replaces the cost model — disjoint keys ≈ free (page-table
binary search); huge key set → cap the list, skip the filter, ≡ hash join;
"should this filter use the index at all" = the scanner's general cost
problem, solved once for all queries (Phase 4b).

---

## Part II — Transition (worked backwards)

### Gap: this branch vs the end state

Branch implements the probe as `IndexProbeTarget`: a second target
TableProvider (per-batch `MapIndexExec` + `DistinctRowAddrsExec` dedup +
hand-rolled unindexed union), selected by pre-planning branches
(`target_provider`/`probe_lookups`). It duplicates the scan stack in
merge-owned code — the exact accretion pattern this doc exists to stop.
Underlying index work is identical either way (both build
`ScalarIndexExpr::IsIn`). **Decision: replace with static filter pushdown;
salvage tests, benches, over/under-match analysis.** Full lib suite is green
at current commit (2629 passed) — safety net for the swap.

### Phase 1: probe = static key-filter pushdown  (delivers priority 1)

1. `with_key_filter` helper (the ONE transitional line in `create_plan`):
   Some iff `source_replayable && by_source == Keep && use_index &&
   key_count ≤ CAP`. Collect distinct non-null source key values at plan time
   (materialized source: already in memory). No index checks, no
   storage-version checks — scanner's job.
2. `CAP` = single named constant (bounded-input guard, not routing; 4b retires
   it). Start ~100k keys; benchmark.
3. Delete `probe.rs`, `target_provider` branch, `probe_lookups`. v1 keeps
   `MapIndexExec` until Phase 3.
4. Routing flip: drop `would_use_scalar_index` from `can_use_create_plan`.
5. Append-downgrade on zero matches (sink runtime; pure-upsert shape only).
6. Verify index query is visible in `explain_plan` output; analyze shows scan
   + filter metrics.
7. Tests: re-point branch plan-shape tests (pushed filter + index query in
   explain; streaming → no filter; delete-by-source → no filter); semantic
   matrix unchanged; UUIDv7 ~zero page IO + Append downgrade; multi-batch
   source; one merge-level integration test for unindexed-fragment coverage;
   multi-fragment throughout.
8. Bench parity gate (`benches/merge_insert.rs`, IO via IoTrackingStore):
   v2-filter vs v1-indexed vs v2-hash-join across ratios / key shapes /
   composite keys. Acceptance: v2-filter ≥ v1-indexed on all shapes.

### Phase 2: RewriteColumns sink  (parallel track; blocks v1 deletion)

v2 rewrites full rows for partial-schema sources — write amplification AND a
semantic regression vs v1 (`RewriteColumns` keeps row addrs stable; indexes on
untouched columns stay valid).

1. Sink strategy = the pure function from Part I §4; planner maps node → exec
   via it, nothing else.
2. `PartialColsMergeInsertExec`: port v1 `update_fragments` largely as-is
   (already stream-based); tombstone superseded field ids; inserts as full-row
   new fragments, same transaction. Preserve semantic restrictions
   (non-nullable missing cols × inserts); lift implementation-artifact ones.
3. Stable-row-id metadata refresh parity.
4. Full metrics + `inserted_rows_filter` from ALL sinks; strategy in explain.
5. Tests: port/parametrize v1 `subcols` module; file-layout assertions;
   interleave vs full-rewrite; multi-fragment; stable row ids on/off.

### Phase 3: delete v1  (delivers priority 2)

1. Flip remaining `can_use_create_plan` gates (indexed fell in 1, subschema
   in 2), each with bench-parity evidence.
2. Delete the Part I inventory's DELETED list (~2500+ lines).
3. explain/analyze unconditional. Retry loop verified per sink;
   `KeyExistenceFilter` wiring unchanged.
4. Bindings sweep: `use_index` documented as "may push key filter" hint.

Exit: bench suite no regression > noise; semantic matrix green on one path.

### Phase 4: shared-stack work  (no merge changes; each independently landable)

- **4a. lance-index IsIn eval**: sort keys once per query; finger/gallop
  cursor over page table (O(N+P)); per-page merge; prefetch next pages —
  main IO-latency win on object storage; skip null pages for non-null probes.
  Prereq: value-ordered in-memory page batch — preserve #5446's wins
  (id-sorted maps from a sorted copy; sort matched ids k≤4096 before
  `from_sorted_iter`; re-run `lance-index/benches/btree.rs`,
  `lance/benches/scalar_index.rs`).
- **4b. Scanner IsIn cost model**: selectivity from page-table overlap (exact
  touched-page upper bound, pre-IO); optionally discount cache-resident pages
  (#7862). Retires Phase 1's CAP.
- **4c. Semi-join index acceleration as shared machinery**: dynamic-filter
  API on FilteredReadExec + a lance-datafusion rule for `Join(build, probe =
  Lance scan)`. Streaming sources get the probe; **delete `with_key_filter`
  from create_plan** — lowering reaches its final form. Any DF join over a
  Lance table benefits, not just merge.
- **4d. Parallel probe partitioning** (page-table range boundaries), only if
  profiles demand.

### Phase 5: overlay write sink  (post-#8016 GA; after Phase 2)

Runtime-adaptive inside `PartialColsMergeInsertExec` (invariant 3 — never a
planner decision): rows already grouped per fragment in the sink; choose
overlay vs column-rewrite per fragment by observed match density (initial
heuristic `touched/fragment_rows < ~5%`, spec TODO `data_overlay_file.md:348`;
tune with #7544 suite).

1. Eligibility in the sink: `overlays_enabled()` AND partial-schema.
2. Transaction shape: updates-as-overlays + inserts-as-fragments in ONE
   commit; `Operation::Update` extension vs composite op — spec PR first.
3. Row-lineage gap: overlay commit arm must bump `last_updated_at_version`.
4. Compaction pressure vs `max_overlays_per_fragment=10`: measure (#7544);
   overlay→overlay merge as follow-on.
5. Tests: end-to-end overlay production; read-back; index staleness masking;
   compaction folds; conflict matrix; crossover unit tests.

---

## Definition of done (measurable)

- One execution path; one plan shape (modulo sink node choice).
- `create_plan` ≤ ~50 lines, zero cost branches, passes the grep test (§6.1).
- explain/analyze work for every merge; all sinks report full metrics.
- Merge-specific non-test LOC ≈ halved; each phase net-negative (Phase 2
  excepted).
- Adding a predicated WHEN clause touches only {params, assign_action} — the
  growth-test table holds.

## Resolved

- Probe architecture: filter pushdown, not a custom provider/exec
  (2026-07-28; `IndexProbeTarget` to be discarded, tests/benches salvaged).
- Sink-selection home: pure shape function + sink runtime adaptivity; no
  optimizer rule (was rev-1 open q5).
- Composite-key intersect: dissolves into conjunct pushdown — scanner's
  problem (was rev-1 open q1).

## Unanswered questions

1. Phase-1 CAP: value, and behavior above it — skip filter entirely vs push
   min/max range only (still prunes clustered keys). Benchmark.
2. Large InList through DF logical optimizer at CAP size — planning-time cost
   acceptable? (Measure; if bad, lower CAP or inject post-optimization.)
3. Does the scanner always choose the index for IsIn when present (no cost
   check)? If yes, over-probing risk at large CAP → informs q1; 4b fixes
   properly.
4. `use_index=false` on v2 = "don't push key filter" — acceptable
   reinterpretation, or deprecate toward a scanner/session knob?
5. Phase 5 transaction shape: extend `Operation::Update` vs composite op —
   needs format owners.
6. `merge_session()`: today create_plan builds a fresh `SessionContext` per
   merge; end state wants lance rules registered once. Shared session vs
   per-merge with a rule-registration helper?
